dipy/reconst/ivim.py

```diff
@@ -0,0 +1,541 @@
+
+        x0 = np.array([S0, f, D_star, D])
+        # Fit parameters again with scaling
+        params = self._leastsq(data, x0)
+
+        if params[1] > self.f_threshold:
+            params = x0
+
+        return IvimFit(self, params)
+
+    def estimate_linear_fit(self, data, split_b, less_than=True):
+        """Estimate a linear fit by taking log of data.
+
+        Parameters
+        ----------
+        data : array
+            An array containing the data to be fit
+
+        split_b : float
+            The b value to split the data
+
+        less_than : bool
+            If True, splitting occurs for bvalues less than split_b
+
+        Returns
+        -------
+        S0 : float
+            The estimated S0 value. (intercept)
+
+        D : float
+            The estimated value of D.
+        """
+        if less_than:
+            bvals_split = self.gtab.bvals[self.gtab.bvals <= split_b]
+            D, neg_log_S0 = np.polyfit(bvals_split,
+                                       -np.log(data[self.gtab.bvals <=
+                                                    split_b]), 1)
+        else:
+            bvals_split = self.gtab.bvals[self.gtab.bvals >= split_b]
+            D, neg_log_S0 = np.polyfit(bvals_split,
+                                       -np.log(data[self.gtab.bvals >=
+                                                    split_b]), 1)
+
+        S0 = np.exp(-neg_log_S0)
+        return S0, D
+
+    def estimate_f_D_star(self, params_f_D, data, S0, D):
+        """Estimate D_star using the values of all the other parameters obtained before
+
+        Parameters
+        ----------
+        params_f_D : array
+            An array containing the value of f and D_star.
+
+        data : array
+            Array containing the actual signal values.
+
+        S0 : float
+            The parameters S0 obtained from a linear fit.
+
+        D : float
+            The parameters D obtained from a linear fit.
+
+        Returns
+        -------
+        f : float
+           Perfusion fraction estimated from the fit.
+        D_star :
+            The value of D_star estimated from the fit.
+        """
+        gtol = self.options["gtol"]
+        ftol = self.options["ftol"]
+        xtol = self.tol
+        epsfcn = self.options["eps"]
+        maxfev = self.options["maxiter"]
+
+        if SCIPY_LESS_0_17:
+            try:
+                res = leastsq(f_D_star_error,
+                              params_f_D,
+                              args=(self.gtab, data, S0, D),
+                              gtol=gtol,
+                              xtol=xtol,
+                              ftol=ftol,
+                              epsfcn=epsfcn,
+                              maxfev=maxfev)
+                f, D_star = res[0]
+                return f, D_star
+            except:
+                warningMsg = "x0 obtained from linear fitting is unfeasibile as "
```

Comment:
I'm confused by this warningMsg. It's if leastsq fails but the warning is that x0 from the linear fitting is not feasible. Is that really true or is that just a common cause of leastsq failing? Maybe preface what you have with "Non-linear fitting using leastsq failed"?